drivers/i2c/i2c_sam0.c

```diff
@@ -0,0 +1,521 @@
+		baud_low = baud - baud_high;
+		if (baud_low < 1 && baud_high > 1) {
+			--baud_high;
+			++baud_low;
+		}
+		if (baud_low < 1 || baud_low > 255) {
+			return -ERANGE;
+		}
+
+#ifdef SERCOM_I2CM_BAUD_HSBAUD
+		LOG_DBG("Setting %s to high speed with divisors %u/%u",
+			DEV_NAME(dev), baud_high, baud_low);
+
+		/*
+		 * 48 is just from the app notes, but the datasheet says
+		 * it's ignored
+		 */
+		i2c->BAUD.reg = SERCOM_I2CM_BAUD_HSBAUD(baud_high) |
+				SERCOM_I2CM_BAUD_HSBAUDLOW(baud_low) |
+				SERCOM_I2CM_BAUD_BAUD(48) |
+				SERCOM_I2CM_BAUD_BAUDLOW(48);
+#else
+		return -ENOTSUP;
+#endif
+
+	default:
+		return -ENOTSUP;
+	}
+
+	wait_synchronization(i2c);
+	return 0;
+}
+
+static int i2c_sam0_configure(struct device *dev, u32_t config)
+{
+	const struct i2c_sam0_dev_config *const cfg = DEV_CFG(dev);
+	SercomI2cm *const i2c = cfg->regs;
+	int retval;
+
+	if (!(config & I2C_MODE_MASTER)) {
+		return -EINVAL;
+	}
+
+	if (config & I2C_SPEED_MASK) {
+		i2c->CTRLA.bit.ENABLE = 0;
+		wait_synchronization(i2c);
+
+		retval = i2c_sam0_set_apply_bitrate(dev, config);
+
+		i2c->CTRLA.bit.ENABLE = 1;
+		wait_synchronization(i2c);
+
+		if (retval != 0) {
+			return retval;
+		}
+	}
+
+	return 0;
+}
+
+static int i2c_sam0_initialize(struct device *dev)
+{
+	struct i2c_sam0_dev_data *data = DEV_DATA(dev);
+	const struct i2c_sam0_dev_config *const cfg = DEV_CFG(dev);
+	SercomI2cm *const i2c = cfg->regs;
+	int retval;
+
+	/* Enable the GCLK */
+	GCLK->CLKCTRL.reg = cfg->gclk_clkctrl_id | GCLK_CLKCTRL_GEN_GCLK0 |
+			GCLK_CLKCTRL_CLKEN;
+
+	/* Enable SERCOM clock in PM */
+	PM->APBCMASK.reg |= cfg->pm_apbcmask;
+
+	/* Disable all I2C interrupts */
+	i2c->INTENCLR.reg = SERCOM_I2CM_INTENCLR_MASK;
+
+	/* I2C mode, enable timeouts */
+	i2c->CTRLA.reg = SERCOM_I2CM_CTRLA_MODE_I2C_MASTER |
+#ifdef SERCOM_I2CM_CTRLA_LOWTOUTEN
+			SERCOM_I2CM_CTRLA_LOWTOUTEN |
+#endif
+			SERCOM_I2CM_CTRLA_INACTOUT(0x3);
+	wait_synchronization(i2c);
+
+	/* Enable smart mode (auto ACK) */
+	i2c->CTRLB.reg = SERCOM_I2CM_CTRLB_SMEN;
+	wait_synchronization(i2c);
+
+	retval = i2c_sam0_set_apply_bitrate(dev,
+			_i2c_map_dt_bitrate(cfg->bitrate));
+	if (retval != 0) {
+		return retval;
+	}
+
+	k_sem_init(&data->sem, 0, 1);
+
+	cfg->irq_config_func(dev);
+
+	i2c->CTRLA.bit.ENABLE = 1;
+	wait_synchronization(i2c);
+
+	/* Force bus idle */
+	i2c->STATUS.bit.BUSSTATE = 1;
+	wait_synchronization(i2c);
+
+	return 0;
+}
+
+
+static const struct i2c_driver_api i2c_sam0_driver_api = {
+	.configure = i2c_sam0_configure,
+	.transfer = i2c_sam0_transfer,
+};
+
+#define I2C_SAM0_DEVICE(n)						      \
+	static void i2c_sam_irq_config_##n(struct device *dev);		      \
+	static const struct i2c_sam0_dev_config i2c_sam0_dev_config_##n = {   \
+		.regs = (SercomI2cm *)DT_ATMEL_SAM0_I2C_##n##_BASE_ADDRESS,   \
+		.bitrate = DT_ATMEL_SAM0_I2C_##n##_CLOCK_FREQUENCY,	      \
+		.pm_apbcmask = PM_APBCMASK_SERCOM##n,			      \
+		.gclk_clkctrl_id = GCLK_CLKCTRL_ID_SERCOM##n##_CORE,	      \
```

Comment:
Yea things break right here as SERCOMn != I2Cn 

I'm afraid we have to keep to using `dts_fixup.h` after all :\

Reply:
Well that can be worked around with something like:

```diff
--- a/drivers/i2c/i2c_sam0.c
+++ b/drivers/i2c/i2c_sam0.c
@@ -464,13 +464,15 @@ static const struct i2c_driver_api i2c_sam0_driver_api = {
        .transfer = i2c_sam0_transfer,
 };
 
+#define I2C_SAM0_CLKID(n) UTIL_CAT(GCLK_CLKCTRL_ID_SERCOM, UTIL_CAT(DT_ATMEL_SAM0_I2C_DT_I2C_ ## n ## _SERCOM, _CORE))
+
 #define I2C_SAM0_DEVICE(n)                                                   \
        static void i2c_sam_irq_config_##n(struct device *dev);               \
        static const struct i2c_sam0_dev_config i2c_sam0_dev_config_##n = {   \
                .regs = (SercomI2cm *)DT_ATMEL_SAM0_I2C_##n##_BASE_ADDRESS,   \
                .bitrate = DT_ATMEL_SAM0_I2C_##n##_CLOCK_FREQUENCY,           \
                .pm_apbcmask = PM_APBCMASK_SERCOM##n,                         \
-               .gclk_clkctrl_id = GCLK_CLKCTRL_ID_SERCOM##n##_CORE,          \
+               .gclk_clkctrl_id = I2C_SAM0_CLKID(n),                         \
                .irq_config_func = &i2c_sam_irq_config_##n                    \
        };                                                                    \
        static struct i2c_sam0_dev_data i2c_sam0_dev_data_##n;                \
diff --git a/dts/arm/atmel/samd.dtsi b/dts/arm/atmel/samd.dtsi
index 555964cc0d..719a8a7efd 100644
--- a/dts/arm/atmel/samd.dtsi
+++ b/dts/arm/atmel/samd.dtsi
@@ -65,6 +65,7 @@
 
                sercom0: sercom@42000800 {
                        compatible = "atmel,sam0-sercom";
+                       sercom = <0>;
                        reg = <0x42000800 0x40>;
                        interrupts = <9 0>;
                        status = "disabled";
@@ -73,6 +74,7 @@
 
                sercom1: sercom@42000c00 {
                        compatible = "atmel,sam0-sercom";
+                       sercom = <1>;
                        reg = <0x42000c00 0x40>;
                        interrupts = <10 0>;
                        status = "disabled";
@@ -81,6 +83,7 @@
 
                sercom2: sercom@42001000 {
                        compatible = "atmel,sam0-sercom";
+                       sercom = <2>;
                        reg = <0x42001000 0x40>;
                        interrupts = <11 0>;
                        status = "disabled";
@@ -89,6 +92,7 @@
 
                sercom3: sercom@42001400 {
                        compatible = "atmel,sam0-sercom";
+                       sercom = <3>;
                        reg = <0x42001400 0x40>;
                        interrupts = <12 0>;
                        status = "disabled";
@@ -97,6 +101,7 @@
 
                sercom4: sercom@42001800 {
                        compatible = "atmel,sam0-sercom";
+                       sercom = <4>;
                        reg = <0x42001800 0x40>;
                        interrupts = <13 0>;
                        status = "disabled";
@@ -105,6 +110,7 @@
 
                sercom5: sercom@42001c00 {
                        compatible = "atmel,sam0-sercom";
+                       sercom = <5>;
                        reg = <0x42001c00 0x40>;
                        interrupts = <14 0>;
                        status = "disabled";
diff --git a/dts/bindings/i2c/atmel,sam0-i2c.yaml b/dts/bindings/i2c/atmel,sam0-i2c.yaml
index 4d14fff29e..e1ca4e2948 100644
--- a/dts/bindings/i2c/atmel,sam0-i2c.yaml
+++ b/dts/bindings/i2c/atmel,sam0-i2c.yaml
@@ -31,4 +31,10 @@ properties:
       category: required
       description: required interrupts
       generation: define
+
+    sercom:
+      type: int
+      category: required
+      description: SERCOM device index
+      generation: define
 ...
```

But that still leaves the problem of getting a stable compile time mapping of "is SERCOM3 enabled as I2C" for pinmux.  Either way, that's probably beyond the scope of what this PR should do, so I'll put the minimal `BASE_ADDRESS` mappings back in `dts_fixup.h` for now.